fix: NetworkManager ApprovalTimeout should not depend upon client synchronization

NoelStephensUnity opened the pull request:
This bug was discovered in a Boss Room play test where the ClientConnectionBufferTimeout was set to 1 second. The underlying issue was that the ApprovalTimeout coroutine would depend upon the NetworkManager.IsConnectedClient to be determine if the client has been approved.  This is problematic because NetworkManager.IsConnectedClient (when scene management is enabled) is set once the client has already been approved **and** has been **fully synchronized**.  If the time it takes a client to synchronize exceeds the ClientConnectionBufferTimeout then while the client was approved it would time out anyway.

**Here is how this fix addresses the issue:** 
- It adds a new internally set property NetworkManager.IsApproved which is set upon the client being approved.
- The timeout messages have been updated with more pertinent/useful information about the timeout.

[MTT-4895](https://jira.unity3d.com/browse/MTT-4895)

## Changelog
- Added: `NetworkManager.IsApproved` flag that is set to `true` a client has been approved.
- Fixed: `NetworkManager.ApprovalTimeout` will not timeout due to slower client synchronization times as it now uses the added `NetworkManager.IsApproved` flag to determined if the client has been approved or not.

## Testing and Documentation
- Includes integration test update to ConnectionApprovalTimeoutTests.
- Includes new API xml documentation for NetworkManager.IsApproved

NoelStephensUnity committed:
fix

This fix separates the IsConnectedClient from the approval process by adding an IsApproved flag.
This also has a possible fix for the domain backup error during serialization.

NoelStephensUnity committed:
fix

This is the final update for the ApprovalTimeout updates.
It accounts for latency and logs more meaningful information for the users.
The latency adjustments are capped at no more than 1 second longer than the specified ClientConnectionBufferTimeout to prevent from latency derived attacks.

NoelStephensUnity committed:
update

Added the ability to bypass the entire NetcodeIntegrationTest connection approval process after the server and clients have been started.  This allows us to now use NetcodeIntegrationTest for unique connection oriented tests without being bound to the asserts if not all clients connected properly.

NoelStephensUnity committed:
test

This is a bit of a refactoring for ConnectionApprovalTimeoutTests. It uses the added m_BypassConnectionTimeout to bypass the waiting for clients to connect. It still uses the message hook catch technique to simulate the timeout scenarios where either a server detects a transport connection but never receives a connection request or a client sends the connection request but never receives approval for the connection.

NoelStephensUnity committed:
Merge branch 'develop' into fix/separate-approval-from-clientconnected

NoelStephensUnity committed:
update

Updating change log

NoelStephensUnity commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -232,6 +232,10 @@ private void BuildDestinationToTransitionInfoTable()
         private void BuildTransitionStateInfoList()
         {
 #if UNITY_EDITOR
+            if (UnityEditor.EditorApplication.isUpdating)
```
This might get removed and placed into a really small PR, but it seems to fix the "domain backup" errors we have been seeing during serialization with the more recent versions of Unity (a known issue).

NoelStephensUnity committed:
style

minor adjustment to the comments.

NoelStephensUnity committed:
fix

Fixing issue where a shutdown could be in progress and the transport driver was already disconnected/deallocated.

NoelStephensUnity committed:
fix

Adding the early exit if a shutdown is in progress

NoelStephensUnity committed:
test update

updating some of the names and comments.

NoelStephensUnity committed:
Merge branch 'develop' into fix/separate-approval-from-clientconnected

jeffreyrainy commented on com.unity.netcode.gameobjects/Runtime/Core/NetworkManager.cs:
```diff
@@ -1649,59 +1665,98 @@ private void SendConnectionRequest()
-
-                //We yield every frame incase a pending client disconnects and someone else gets its connection id
-                while (IsListening && (LocalTime - timeStarted).Time < NetworkConfig.ClientConnectionBufferTimeout && PendingClients.ContainsKey(clientId))
+            // Use the current RTT latency between the client and server (instance relative)
```
In my opinion, the proper way would be to take a configurable timeout from the game team and not adjust for latency.

If we fear the gamedevs will forget about latency or be surprised, we can add a fixed amount, say 0.25s. But using computed RTT adds unnecessary complexity and mixes abstractions levels.

NoelStephensUnity committed:
update

Removing the RTT latency stuff.

NoelStephensUnity committed:
test

was not resetting the m_BypassConnectionTimeout value... although this shouldn't have mattered that much since it is reset for each new NetcodeIntegrationTest instance.
